vector/src/main/java/im/vector/riotx/features/crypto/keysbackup/restore/KeysBackupRestoreSharedViewModel.kt

```diff
@@ -62,30 +84,187 @@ class KeysBackupRestoreSharedViewModel @Inject constructor() : ViewModel() {
+                    // we need to restore directly
+                    keySourceModel.postValue(
+                            KeySource(isInMemory = false, isInQuadS = false)
+                    )
+                }
+
+                loadingEvent.postValue(null)
+            } catch (failure: Throwable) {
+                loadingEvent.postValue(null)
+                _keyVersionResultError.postValue(LiveEvent(stringProvider.getString(R.string.keys_backup_get_version_error, failure.localizedMessage)))
+            }
+        }
+    }
+
+    fun handleGotSecretFromSSSS(cipherData: String, alias: String) {
+        try {
+            cipherData.fromBase64().inputStream().use { ins ->
+                val res = session.loadSecureSecret<Map<String, String>>(ins, alias)
+                val secret = res?.get(KEYBACKUP_SECRET_SSSS_NAME)
+                if (secret == null) {
+                    _navigateEvent.postValue(
+                            LiveEvent(NAVIGATE_FAILED_TO_LOAD_4S)
+                    )
+                    return
+                }
+                loadingEvent.value = WaitingViewData(stringProvider.getString(R.string.keys_backup_restore_is_getting_backup_version))
+
+                viewModelScope.launch(Dispatchers.IO) {
+                    try {
+                        recoverUsingBackupRecoveryKey(computeRecoveryKey(secret.fromBase64()))
+                    } catch (failure: Throwable) {
+                        _navigateEvent.postValue(
+                                LiveEvent(NAVIGATE_FAILED_TO_LOAD_4S)
+                        )
+                    }
                 }
             }
+        } catch (failure: Throwable) {
+        }
+    }
+
+    suspend fun recoverUsingBackupPass(passphrase: String) {
+        val keysBackup = session.cryptoService().keysBackupService()
+        val keyVersion = keyVersionResult.value ?: return
 
-            override fun onFailure(failure: Throwable) {
-                loadingEvent.value = null
-                _keyVersionResultError.value = LiveEvent(context.getString(R.string.keys_backup_get_version_error, failure.localizedMessage))
+        loadingEvent.postValue(WaitingViewData(stringProvider.getString(R.string.loading)))
 
-                // TODO For network error
-                //  _keyVersionResultError.value = LiveEvent(context.getString(R.string.network_error_please_check_and_retry))
+        try {
+            val result = awaitCallback<ImportRoomKeysResult> {
+                keysBackup.restoreKeyBackupWithPassword(keyVersion,
+                        passphrase,
+                        null,
+                        session.myUserId,
+                        progressObserver,
+                        it
+                )
             }
-        })
+            loadingEvent.postValue(null)
+            didRecoverSucceed(result)
+            trustOnDecrypt(keysBackup, keyVersion)
+        } catch (failure: Throwable) {
+            throw failure
```

Comment:
the try catch seems a bit useless here, no?

Reply:
good catch, I should  loadingEvent.postValue(null) in the catch. thx